### PR TITLE
Feature/add sklearn e2e tests

### PR DESCRIPTION
**Pull request by srcansiz**
**PR description**
- Solves the bug regarding  not failing once a process finishes 
- Solves bug un-serializable object MetricTypes for experiment breakpoint save state and load stat. 
- Implements e2e tests for sklearn

**Developer Certificate Of Origin (DCO)**

By opening this merge request, you agree to the
[Developer Certificate of Origin (DCO)](https://github.com/fedbiomed/fedbiomed/blob/master/CONTRIBUTING.md#fed-biomed-developer-certificate-of-origin-dco)

This DCO essentially means that:

- you offer the changes under the same license agreement as the project, and
- you have the right to do that,
- you did not steal somebody else’s work.

**License**

Project code files should begin with these comment lines to help trace their origin:
```
# This file is originally part of Fed-BioMed
# SPDX-License-Identifier: Apache-2.0
```

Code files can be reused from another project with a compatible non-contaminating license.
They shall retain the original license and copyright mentions.
The `CREDIT.md` file and `credit/` directory shall be completed and updated accordingly.

**Guidelines for PR review**

General:

* give a glance to [DoD](https://fedbiomed.org/latest/developer/definition-of-done/)
* check [coding rules and coding style](https://fedbiomed.org/latest/developer/usage_and_tools/#coding-style)
* check docstrings (eg run `tests/docstrings/check_docstrings`)

Specific to some cases:

* update all conda envs consistently (`development` and `vpn`, Linux and MacOS)
* if modified researcher (eg new attributes in classes) check if breakpoint needs update (`breakpoint`/`load_breakpoint` in `Experiment()`, `save_state_breakpoint`/`load_state_breakpoint` in aggregators, strategies, secagg, etc.)
* if modified a component with versioning (config files, breakpoint, messaging protocol) then update the version following the rules in `common/utils/_versions.py`

### CODE REVIEW

**Comment by ybouilla** [comment]
Hey Sergen, I see you updated `TrainingArgs` class  with a `load_state_breakpoint` method: i can see the utility (to check parameters are loaded correctly). Is there another purpose here?

**Comment by srcansiz** [comment]
> Hey Sergen, I see you updated `TrainingArgs` class with a `load_state_breakpoint` method: i can see the utility (to check parameters are loaded correctly). Is there another purpose here?

Hi @ybouilla, 

I updated `TrainingArgs` to solve the bugs in breakpoints. It was raising JSON serialization error due to `MetricTypes`. I encountered this error while testing  an experiment with breakpoints enabled.  Threrefore, I implemented to method `get_state_brekpoint` that only returns name of the metric. However, while loding the breakpoint string type metric should be converted to MetricTypes, otherwise, node raises error due to unrecognized metric type. So, this is why I added another method `load_state_brekpoint`. This is a simple solution to solve this bug but I think we need to have a unique way of saving such objects in break points.

File: tests/end2end/experiments/training_plans/sklearn.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        NUMBER_COLS = 20
+        dataset = pd.read_csv(self.dataset_path,header=None,delimiter=',')
+        X = dataset.iloc[:,0:NUMBER_COLS].values
+        y = dataset.iloc[:,NUMBER_COLS]
+        return DataManager(dataset=X,target=y.values, shuffle=True)
+
+
+class SGDRegressorTrainingPlan(FedSGDRegressor):
+    def training_data(self):
+        dataset = pd.read_csv(self.dataset_path, delimiter=';', header=0)
```
Comment (ybouilla) [inline tests/end2end/experiments/training_plans/sklearn.py:24]: should it be `delimiter=';'` here instead? Like you did for the other training plans

**Review by ybouilla — COMMENTED** [review]
Thanks for this PR:
Here are some questions and comments about your work.

Regarding the tests, it could be nice if you add the following tests:
- testing TrainingPlan with several `test_metrics`
- testing sklearn Model with native `Scaffold`
- testing sklearn with declearn's `Scaffold` / fedopt

**Comment by ybouilla** [comment]
> I updated TrainingArgs to solve the bugs in breakpoints. It was raising JSON serialization error due to MetricTypes. I encountered this error while testing an experiment with breakpoints enabled. Threrefore, I implemented to method get_state_brekpoint that only returns name of the metric. However, while loding the breakpoint string type metric should be converted to MetricTypes, otherwise, node raises error due to unrecognized metric type. So, this is why I added another method load_state_brekpoint. This is a simple solution to solve this bug but I think we need to have a unique way of saving such objects in break points.

Ok I did not know it was due to this bug. It has never been reported before. Congrats to have spotted it!

File: tests/end2end/e2e_sklearn.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+                     aggregator=FedAverage(),
+                     node_selection_strategy=None)
+
+    exp.run()
+
+    clear_experiment_data(exp)
+
+
+
+def test_06_seklearn_adni_regressor_with_secureaggregation():
```
[the remark this answers is not in the document]
Comment (srcansiz) [reply tests/end2end/e2e_sklearn.py:360]: Hi @ybouilla, 

Good point. I couldn't decide where to test it. I wanted to leave it here just to test if sklearn model parameters are compatible with secure aggregation since all the dataset training plan etc was already defined in this test file.

File: tests/end2end/experiments/training_plans/sklearn.py (hunk 1 above)
Comment (srcansiz) [reply tests/end2end/experiments/training_plans/sklearn.py:24]: In the dataset that is saved in the notebook directory the `delimeter` is `;`

File: tests/end2end/e2e_sklearn.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+        time.sleep(5)
+
+
+        # Run tests
+        yield node_1, node_2, node_3, researcher
+
+
+        kill_subprocesses(node_processes)
+        thread.join()
+
```
[the remark this answers is not in the document]
Comment (srcansiz) [reply tests/end2end/e2e_sklearn.py:120]: Hi @ybouilla,

I added this but it is not mandatory because once all the node processes are killed the thread will finish. However, it was just to explicitly show that that is a thread running to follow node processes. I can add it to other tests as well.

File: tests/end2end/e2e_sklearn.py [hunk 4]
```diff
@@ -0,0 +1,350 @@
...
+    """Generate testing data"""
+
+    path = get_data_folder('sklearn-e2e-tests')
+    p1 = os.path.join(path , 'c1.csv')
+    p2 = os.path.join(path , 'c2.csv')
+    p3 = os.path.join(path , 'c3.csv')
+
+    # If there is path stop generating the data again
+    if os.path.isdir(path):
+        return p1, p2, p3
```
[the remark this answers is not in the document]
Comment (srcansiz) [reply tests/end2end/e2e_sklearn.py:136]: Hi @ybouilla,

Thanks you are right and it was the reason that you were having error in the first place while reviewing this PR. So therefore, it is already fixed.

**Comment by srcansiz** [comment]
Hi @ybouilla, 

I implemented your suggestions. Could you please control and let me know if it is okay? 

Thank you

File: tests/end2end/experiments/training_plans/sklearn.py (hunk 1 above)
Comment (ybouilla) [reply tests/end2end/experiments/training_plans/sklearn.py:24]: Oh ok sorry adni dataset is indeed with `;`, whereas the one generated are with `,`, hence the confusion
